airflow/providers/pinecone/hooks/pinecone_hook.py

```diff
@@ -0,0 +1,122 @@
+        UpsertResponse,
+    )
+
+
+class PineconeHook(BaseHook):
+    """
+    Interact with Pinecone. This hook uses the Pinecone conn_id.
+
+    :param conn_id: The connection id to use when connecting to Pinecone.
+    """
+
+    conn_name_attr = "conn_id"
+    default_conn_name = "pinecone_default"
+    conn_type = "pinecone"
+    hook_name = "Pinecone"
+
+    @classmethod
+    def get_ui_field_behaviour(cls) -> dict[str, Any]:
+        """Returns custom field behaviour."""
+        return {
+            "hidden_fields": ["port"],
+            "relabeling": {
+                "login": "Pinecone Environment",
+                "password": "Pinecone API key",
+                "schema": "Project ID",
+            },
+            "placeholders": {
+                "login": "Pinecone Environment",
+                "password": "Pinecone API key",
+                "schema": "Project ID",
+            },
+        }
+
+    def __init__(self, conn_id: str = default_conn_name) -> None:
+        self.conn_id = conn_id
+        self.get_conn()
+
+    def get_conn(self) -> None:
+        pinecone_connection = self.get_connection(self.conn_id)
+        api_key = pinecone_connection.password
+        pinecone_environment = pinecone_connection.login
+        pinecone_host = pinecone_connection.host
+        pinecone_project_name = pinecone_connection.schema
+        log_level = pinecone_connection.extra_dejson.get("log_level", "~/.pinecone")
+        pinecone.init(
+            api_key=api_key,
+            environment=pinecone_environment,
+            host=pinecone_host,
+            project_name=pinecone_project_name,
+            log_level=log_level,
+        )
+
+    def test_connection(self) -> tuple[bool, str]:
+        try:
+            pinecone.list_indexes()
+            return True, "Connection established"
+        except Exception as e:
+            return False, str(e)
+
+    @staticmethod
+    def upsert(
+        index_name: str,
+        vectors: list[Any],
+        namespace: str = "",
+        batch_size: int | None = None,
+        show_progress: bool = True,
+        **kwargs: Any,
+    ) -> UpsertResponse:
+        """
+        The upsert operation writes vectors into a namespace.
+
+        If a new value is upserted for an existing vector id, it will overwrite the previous value.
+
+         .. seealso:: https://docs.pinecone.io/reference/upsert
+
+        To upsert in parallel follow
+
+        .. seealso:: https://docs.pinecone.io/docs/insert-data#sending-upserts-in-parallel
+
+        :param index_name: The name of the index to describe.
+        :param vectors:: A list of vectors to upsert.
+        :param namespace: The namespace to write to. If not specified, the default namespace - "" is used.
+        :param batch_size: The number of vectors to upsert in each batch.
+        :param show_progress: Whether to show a progress bar using tqdm. Applied only
+         if batch_size is provided.
```

Comment:
```suggestion
        :param show_progress: Whether to show a progress bar using tqdm. Applied only
            if batch_size is provided.
```
Just a Sphinx nit for formatting. The second line should be a full hanging indent. Otherwise the Airflow docs for this method will look funky. (Or maybe this can all fit on a single line?)

Reply:
No strong opinion here, but maybe it's worth adding a warning/log message that progress won't be shown if `batch_size` is not provided? And/Or should `show_progress` have a default of False and set to True if `batch_size` is provided?